### PR TITLE
Feature/issue 243 array versions value of

### PR DESCRIPTION
**Pull request by randommm**
#### Summary:
Create versions of value_of and value_of_rec that work for std::vector.

Fix issue #243.

#### Intended Effect:
value_of: convert std::vector<T> to std::vector<double>.
value_of_rec: convert std::vector<T> to std::vector<child_type<T>::type>.

#### How to Verify:
Code review.

#### Side Effects:
None.

#### Documentation:
Included in the new functions.

#### Reviewer Suggestions: 
Anyone.

### CODE REVIEW

**Comment by syclik** [comment]
@rtrangucci, if you have some time, could you look at this pull request?

**Comment by rtrangucci** [comment]
of course, I'll take a look right now.


Rob

**Comment by rtrangucci** [comment]
@randommm can we split this into two PRs? I don't think we should touch the prim/mat/value_of.hpp in this PR. The changes look good to that file, but we should rewrite prim/mat/value_of_rec.hpp the same way.

**Comment by rtrangucci** [comment]
@randommm I should add that I'm happy to create a new issue and tackle the code. Your new prim/mat/value_of function looks to be about 2X as fast as the current implementation.

**Comment by randommm** [comment]
> @randommm <https://github.com/randommm> can we split this into two 
> PRs? I don't think we should touch the prim/mat/value_of.hpp in this 
> PR. The changes look good to that file, but we should rewrite 
> prim/mat/value_of_rec.hpp the same way.
>
Sure, let's split this into another issue and PR.

**Comment by rtrangucci** [comment]
@randommm ok, cool. Do you want me to create the issue or are you already working on it? Again happy to do it, just don't want to get in your way!

**Comment by randommm** [comment]
> @randommm <https://github.com/randommm> I should add that I'm happy to 
> create a new issue and tackle the code. Your new prim/mat/value_of 
> function looks to be about 2X as fast as the current implementation.
>

Have you done some benchmarks? I've done some a while ago: 
https://stackoverflow.com/questions/16283000/most-efficient-way-to-loop-through-an-eigen-matrix/16287999#16287999 
(second answer)




> @randommm <https://github.com/randommm> ok, cool. Do you want me to 
> create the issue or are you already working on it? Again happy to do 
> it, just don't want to get in your way!
>
Don't worry, I'll create an issue and pull request for 
prim/mat/fun/value_of.hpp and prim/mat/fun/value_of_rec.hpp.


However, there's a lot of other places in Stan code where loops in Eigen 
matrices could be done with data pointers too, so probably this is worth 
an issue of its own.

**Comment by bob-carpenter** [comment]
It's very clearly indicated in their doc.  Matrices
are column major.  I'm not sure what else is going on
here, but you can do it with a single loop (even faster)
with:

```
for (int i = 0; i < m.size(); ++i)
  ... do something with m[i] ...
```

That'll iterate in whatever the storage order is, which is
column-major by default.

**Comment by randommm** [comment]
On 02/03/2016 16:49, Bob Carpenter wrote:
> It's very clearly indicated in their doc. Matrices
> are column major. I'm not sure what else is going on
> here, but you can do it with a single loop (even faster)
> with:
>
> ```
> for (int i = 0; i < m.size(); ++i)
> ... do something with m[i] ...
> ```
If m is a matrix, won't this cause an error 
THE_BRACKET_OPERATOR_IS_ONLY_FOR_VECTORS__USE_THE_PARENTHESIS_OPERATOR_INSTEAD
?

What I've done is similar, but first I created pointer using data():

```
const T * data_m = m.data();
for (int i = 0; i < m.size(); ++i)
... do something with data_m[i] ...
```

**Comment by rtrangucci** [comment]
@randommm this all looks good. The only tests we're missing are the mix version of these. @syclik is it worth it to write the mix versions of the value_of and value_of_rec tests?

**Comment by randommm** [comment]
You mean the tests that mix fvar's and var's? They were added together 
with the fwd test, but now I see there's  actually a separated folder 
for that!

**Comment by syclik** [comment]
Yes, please separate out the tests. And we do want to test all the way
through.

File: test/unit/math/fwd/arr/fun/value_of_rec_test.cpp [hunk 1]
```diff
@@ -0,0 +1,62 @@
+#include <stan/math/fwd/core.hpp>
```
Comment (syclik) [inline test/unit/math/fwd/arr/fun/value_of_rec_test.cpp]: only include:
```
#include <stan/math/fwd/arr.hpp>
#include <gtest/gtest.h>
```
for these tests. (this is new and will make testing easier!)

File: test/unit/math/fwd/arr/fun/value_of_test.cpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
+#include <stan/math/fwd/core.hpp>
```
Comment (syclik) [inline test/unit/math/fwd/arr/fun/value_of_test.cpp]: include:
```
#include <stan/math/fwd/arr.hpp>
#include <gtest/gtest.h>
```

File: test/unit/math/mix/arr/fun/value_of_rec_test.cpp [hunk 3]
```diff
@@ -0,0 +1,45 @@
+#include <stan/math/fwd/core.hpp>
```
Comment (syclik) [inline test/unit/math/mix/arr/fun/value_of_rec_test.cpp]: include:
```
#include <stan/math/mix/arr.hpp>
#include <gtest/gtest.h>
```

File: test/unit/math/mix/arr/fun/value_of_test.cpp [hunk 4]
```diff
@@ -0,0 +1,45 @@
+#include <stan/math/fwd/core.hpp>
```
Comment (syclik) [inline test/unit/math/mix/arr/fun/value_of_test.cpp]: include:
```
#include <stan/math/mix/arr.hpp>
#include <gtest/gtest.h>
```

File: test/unit/math/prim/arr/fun/value_of_rec_test.cpp [hunk 5]
```diff
@@ -0,0 +1,39 @@
+#include <stan/math/prim/arr/fun/value_of_rec.hpp>
```
Comment (syclik) [inline test/unit/math/prim/arr/fun/value_of_rec_test.cpp]: include:
```
#include <stan/math/prim/arr.hpp>
#include <gtest/gtest.h>
```

File: test/unit/math/prim/arr/fun/value_of_test.cpp [hunk 6]
```diff
@@ -0,0 +1,39 @@
+#include <stan/math/prim/arr/fun/value_of.hpp>
```
Comment (syclik) [inline test/unit/math/prim/arr/fun/value_of_test.cpp]: include:
```
#include <stan/math/prim/arr.hpp>
#include <gtest/gtest.h>
```

File: test/unit/math/rev/arr/fun/value_of_rec_test.cpp [hunk 7]
```diff
@@ -0,0 +1,43 @@
+#include <stan/math/rev/core.hpp>
```
Comment (syclik) [inline test/unit/math/rev/arr/fun/value_of_rec_test.cpp]: include:
```
#include <stan/math/rev/arr.hpp>
#include <gtest/gtest.h>
```

File: test/unit/math/rev/arr/fun/value_of_test.cpp [hunk 8]
```diff
@@ -0,0 +1,43 @@
+#include <stan/math/rev/core.hpp>
```
Comment (syclik) [inline test/unit/math/rev/arr/fun/value_of_test.cpp]: include:
```
#include <stan/math/rev/arr.hpp>
#include <gtest/gtest.h>
```

**Comment by syclik** [comment]
@randommm, we just added headers to include within the tests to make life easier on all of us. We're going to only include those in our tests from here on out.

Want me to change them? I just wanted you to be aware of it (and mark them down as comments).

**Comment by randommm** [comment]
Thanks, I think they are correct now.

File: stan/math/prim/arr.hpp [hunk 9]
```diff
@@ -14,7 +14,8 @@
 #include <stan/math/prim/arr/fun/rep_array.hpp>
 #include <stan/math/prim/arr/fun/scaled_add.hpp>
 #include <stan/math/prim/arr/fun/sub.hpp>
-#include <stan/math/prim/arr/fun/sum.hpp>
+#include <stan/math/prim/arr/fun/value_of.hpp>
```
Comment (syclik) [inline stan/math/prim/arr.hpp]: we still need to include `sum.hpp`

Comment (randommm) [inline stan/math/prim/arr.hpp]: ops, sorry, removed that line by mistake.

Comment (syclik) [inline stan/math/prim/arr.hpp]: No prob... I already fixed it.

File: stan/math/prim/arr/fun/value_of.hpp [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+#include <stan/math/prim/scal/fun/value_of.hpp>
+#include <stan/math/prim/scal/meta/child_type.hpp>
+#include <vector>
+#include <cstddef>
+
+namespace stan {
+  namespace math {
+
+    /**
+     * Convert a std::vector of type T to a std::vector of doubles.
```
Comment (syclik) [inline stan/math/prim/arr/fun/value_of.hpp]: fix comment to reflect that it'll return a vector of `child_type<T>::type`, not double

File: stan/math/prim/arr/fun/value_of.hpp [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+     *
+     * T must implement value_of. See
```
Comment (syclik) [inline stan/math/prim/arr/fun/value_of.hpp]: remove this comment.

File: test/unit/math/fwd/arr/fun/value_of_rec_test.cpp [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+  x.assign(contents.size(), T());
+  for (size_t i = 0; i < contents.size(); ++i)
+    x[i] = T(contents[i]);
+}
+
+TEST(MathMatrix,value_of_rec) {
+  using stan::math::value_of_rec;
+  using std::vector;
+  using stan::math::fvar;
+  using stan::math::var;
```
Comment (syclik) [inline test/unit/math/fwd/arr/fun/value_of_rec_test.cpp]: remove this line... refers to var in fwd test.

File: test/unit/math/fwd/arr/fun/value_of_rec_test.cpp [hunk 13]
```diff
@@ -0,0 +1,60 @@
+#include <stan/math/fwd/arr.hpp>
+#include <gtest/gtest.h>
+
+template <typename T>
```
Comment (syclik) [inline test/unit/math/fwd/arr/fun/value_of_rec_test.cpp]: use stan/math/prim/arr/fun/fill.hpp (once the other pull request goes in).

File: test/unit/math/fwd/arr/fun/value_of_test.cpp [hunk 14]
```diff
@@ -0,0 +1,60 @@
+#include <stan/math/fwd/arr.hpp>
+#include <gtest/gtest.h>
+
+template <typename T>
```
Comment (syclik) [inline test/unit/math/fwd/arr/fun/value_of_test.cpp]: use stan/math/prim/arr/fun/fill.hpp (once the other pull request goes in).

File: test/unit/math/fwd/arr/fun/value_of_test.cpp [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+  x.assign(contents.size(), T());
+  for (size_t i = 0; i < contents.size(); ++i)
+    x[i] = T(contents[i]);
+}
+
+TEST(MathMatrix,value_of) {
+  using stan::math::value_of;
+  using std::vector;
+  using stan::math::fvar;
+  using stan::math::var;
```
Comment (syclik) [inline test/unit/math/fwd/arr/fun/value_of_test.cpp]: remove use of var

File: test/unit/math/mix/arr/fun/value_of_rec_test.cpp [hunk 16]
```diff
@@ -0,0 +1,42 @@
+#include <stan/math/mix/arr.hpp>
+#include <gtest/gtest.h>
+
+template <typename T>
```
Comment (syclik) [inline test/unit/math/mix/arr/fun/value_of_rec_test.cpp]: use stan/math/prim/arr/fun/fill.hpp (once the other pull request goes in).

File: test/unit/math/mix/arr/fun/value_of_rec_test.cpp [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+  x.assign(contents.size(), T());
+  for (size_t i = 0; i < contents.size(); ++i)
+    x[i] = T(contents[i]);
+}
+
+TEST(MathMatrix,value_of_rec) {
+  using stan::math::value_of_rec;
+  using std::vector;
+  using stan::math::fvar;
+  using stan::math::var;
```
Comment (syclik) [inline test/unit/math/mix/arr/fun/value_of_rec_test.cpp]: remove var from fwd test

File: test/unit/math/mix/arr/fun/value_of_test.cpp [hunk 18]
```diff
@@ -0,0 +1,42 @@
+#include <stan/math/mix/arr.hpp>
+#include <gtest/gtest.h>
+
+template <typename T>
+void fill(const std::vector<double>& contents,
```
Comment (syclik) [inline test/unit/math/mix/arr/fun/value_of_test.cpp]: use stan/math/prim/arr/fun/fill.hpp (once the other pull request goes in).

File: test/unit/math/prim/arr/fun/value_of_rec_test.cpp [hunk 19]
```diff
@@ -0,0 +1,39 @@
+#include <stan/math/prim/arr.hpp>
+#include <gtest/gtest.h>
+
+template <typename T>
+void fill(const std::vector<double>& contents,
```
Comment (syclik) [inline test/unit/math/prim/arr/fun/value_of_rec_test.cpp]: use stan/math/prim/arr/fun/fill.hpp (once the other pull request goes in).

File: test/unit/math/rev/arr/fun/value_of_rec_test.cpp [hunk 20]
```diff
@@ -0,0 +1,40 @@
+#include <stan/math/rev/arr.hpp>
+#include <gtest/gtest.h>
+
+template <typename T>
+void fill(const std::vector<double>& contents,
```
Comment (syclik) [inline test/unit/math/rev/arr/fun/value_of_rec_test.cpp]: use stan/math/prim/arr/fun/fill.hpp (once the other pull request goes in).

File: test/unit/math/rev/arr/fun/value_of_test.cpp [hunk 21]
```diff
@@ -0,0 +1,40 @@
+#include <stan/math/rev/arr.hpp>
+#include <gtest/gtest.h>
+
+template <typename T>
+void fill(const std::vector<double>& contents,
```
Comment (syclik) [inline test/unit/math/rev/arr/fun/value_of_test.cpp]: use stan/math/prim/arr/fun/fill.hpp (once the other pull request goes in).

**Comment by syclik** [comment]
@randommm, looks good. I'll fix the comments I've made, unless you want to.

**Comment by randommm** [comment]
OK, thanks!

File: test/unit/math/fwd/arr/fun/value_of_rec_test.cpp [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+
+  for (size_t i = 0; i < 10; ++i)
+    a_vals.push_back(i + 1);
+
+  vector<double> b_vals;
+
+  for (size_t i = 10; i < 15; ++i)
+    b_vals.push_back(i + 1);
+
+  {
```
Comment (bob-carpenter) [inline test/unit/math/fwd/arr/fun/value_of_rec_test.cpp]: don't need these scope wrappers in C++

**Comment by bob-carpenter** [comment]
@radommm There are still a couple things to take care of in this pull request for includes and some scoping, but it's very near to being done.

**Comment by syclik** [comment]
Jenkins, retest this please.

**Pull request merged**